raiden/transfer/channel.py

```diff
@@ -0,0 +1,1194 @@
+            channel_state.settle_transaction = TransactionExecutionStatus(
+                state_change.block_number,
+                None,
+                None
+            )
+            event = ContractSendChannelSettle(channel_state.identifier)
+            events.append(event)
+
+    return TransitionResult(channel_state, events)
+
+
+def handle_channel_closed(channel_state, state_change):
+    events = list()
+
+    just_closed = (
+        state_change.channel_identifier == channel_state.identifier and
+        get_status(channel_state) in CHANNEL_STATES_PRIOR_TO_CLOSED
+    )
+
+    if just_closed:
+        set_closed(channel_state, state_change.closed_block_number)
+
+        balance_proof = channel_state.partner_state.balance_proof
+        call_update = (
+            state_change.closing_address != channel_state.our_state.address and
+            balance_proof
+        )
+        if call_update:
+            # The channel was closed by our partner, if there is a balance
+            # proof available update this node half of the state
+            update = ContractSendChannelUpdateTransfer(
+                channel_state.identifier,
+                balance_proof,
+            )
+            events.append(update)
+
+        unlock_proofs = get_known_unlocks(channel_state.partner_state)
+        if unlock_proofs:
+            withdraw = ContractSendChannelWithdraw(
+                channel_state.identifier,
+                unlock_proofs,
+            )
+            events.append(withdraw)
+
+    return TransitionResult(channel_state, events)
+
+
+def handle_channel_settled(channel_state, state_change):
+    events = list()
+
+    if state_change.channel_identifier == channel_state.identifier:
+        set_settled(channel_state, state_change.settle_block_number)
+
+    return TransitionResult(channel_state, events)
+
+
+def handle_channel_newbalance(channel_state, state_change):
+    events = list()
+    participant_address = state_change.participant_address
+
+    if participant_address == channel_state.our_state.address:
+        new_balance = max(
+            channel_state.our_state.contract_balance,
+            state_change.contract_balance,
+        )
+        channel_state.our_state.contract_balance = new_balance
+
+    elif participant_address == channel_state.partner_state.address:
+        new_balance = max(
+            channel_state.partner_state.contract_balance,
+            state_change.contract_balance,
+        )
+        channel_state.partner_state.contract_balance = new_balance
+
+    return TransitionResult(channel_state, events)
+
+
+def handle_channel_withdraw(channel_state, state_change):
+    hashlock = state_change.hashlock
+    secret = state_change.secret
+
+    our_withdraw = (
+        state_change.receiver == channel_state.our_state.address and
+        is_locked(channel_state.partner_state, hashlock)
+    )
+    # FIXME: must not remove the lock, otherwise a new unlock proof cannot be
```

Comment:
With the current implementation only the first batch of on-chain unlock will work, because once the unlock is confirmed the merkletree is changed and subsequent merkleproofs will be rejected (the locksroot of the balance proof onchain wont match).

The proper fix is to add another type to represent the state "unlocked on-chain", this type is needed because the merkle tree must be kept intact *and* we must not try to withdraw the same lock twice.

Reply:
Aha! I see. So the only thing to keep in mind here then would be to either:
1. You have it already implemented in your branch (?)
or
2. Make an issue to not forget it. It's easy to forget old FIXMEs in the code.

Reply:
#1281